### ISSUE TITLE
Mypy error when passing None to nn.AdaptiveAvgPool2d or nn.AdaptiveMaxPool2d

### ISSUE THREAD

**Issue by VinhLoiIT** [pytorch/pytorch#49918]
## 🐛 Bug
`nn.AdaptiveAvgPool2d` and `nn.AdaptiveMaxPool2d` support passing `None` to keep the input dim after pooling as shown in the docs like [nn.AdaptiveAvgPool2d](https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveAvgPool2d.html) and [nn.AdaptiveMaxPool2d](https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveMaxPool2d.html#torch.nn.AdaptiveMaxPool2d). However, `mypy` shows error that it is incompatible type when passing `None`.

## To Reproduce

```python
class Model(nn.Module):
    def __init__(self):
        ...
        self.pool = nn.AdaptiveAvgPool2d((1, None))

    def forward(self, inputs):
        # inputs: [B, C, H, W]
        outputs = self.pool(inputs)
        # outputs: [B, C, 1, W]
        return outputs
```
Mypy Error: ```Argument 1 to "AdaptiveAvgPool2d" has incompatible type "Tuple[int, None]"; expected "Union[int, Tuple[int, ...]]"```

## Expected behavior

No error either `nn.AdaptiveAvgPool2d((1, None))` or `nn.AdaptiveAvgPool2d((None, 1))`

## Environment

 - PyTorch Version: 1.7
 - OS: Ubuntu 20.04
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.8.5

cc @albanD @mruberry @jbschlosser @ezyang @malfet @rgommers @xuzhao9 @gramster

**Comment by rgommers** [comment]
Thanks @VinhLoiIT, that's indeed a bug. Looks like it only affects `AdaptiveAvgPool2d`, `AdaptiveAvgPool3d`, `AdaptiveMaxPool2d` and `AdaptiveMaxPool3d`.

I checked that there are tests for using `None`, they're the ones with `desc='tuple_none',` in `torch/testing/_internal/common_nn.py`. Running `mypy` over `test_nn.py` wouldn't have caught this though I think, because of the way test cases are generated.

**Comment by VinhLoiIT** [comment]
When I look at the common types defined in file `torch/nn/common_types.py`, these sizes used in Pytorch are required to be `int` or `Tuple[int, int]`. There is no definition for sizes that have an optional number like the one above (i.e. `(1, None)` or `(None, 1)`). 

Therefore, to keep things concise, I think we could introduce a new types, like `_size_n_opt_t` (e.g. `_size_2_opt_t`) and place it where we accept the optional size like such cases.

So, I suggest adding to `common_types.py` file:
```python
_size_any_opt_t = _scalar_or_tuple_any_t[Optional[int]]
_size_1_opt_t = _scalar_or_tuple_1_t[Optional[int]]
_size_2_opt_t = _scalar_or_tuple_2_t[Optional[int]]
_size_3_opt_t = _scalar_or_tuple_3_t[Optional[int]]
_size_4_opt_t = _scalar_or_tuple_4_t[Optional[int]]
_size_5_opt_t = _scalar_or_tuple_5_t[Optional[int]]
_size_6_opt_t = _scalar_or_tuple_6_t[Optional[int]]
```

So that we can annotate the type for `_MaxPoolNd` (and similar classes as well) like this:
```python
class _MaxPoolNd(Module):
    ...

    def __init__(self, kernel_size: _size_any_opt_t, stride: Optional[_size_any_t] = None,
                 padding: _size_any_t = 0, dilation: _size_any_t = 1,
                 return_indices: bool = False, ceil_mode: bool = False) -> None:
    ...
```

Any thoughts?

**Comment by rgommers** [comment]
`_size_any_opt_t` naming and putting it in `common_types.py` sounds good. I would prefer to not define types that aren't used at all though, I think we need only `_size_any_opt_t`, `_size_2_opt_t` and `_size_3_opt_t`.
